### PR TITLE
Add support for partial burns

### PR DESCRIPTION
**Pull request by thegostep**
NOTE: It will be functionally equivalent to call releaseStake with `_burnAmt = originalStakeAmount` and `_successful = false` to calling `destroyStake()`

Invariants violated by this update:
- StakeReleased event will emit the amount of nmr burned instead of etherReward
- StakeDestroyed event will only emit for full stake burns

### CODE REVIEW

**Comment by kumikoda** [comment]
@thegostep and I were chatting in person: @BryantLuu brought up a good point that we have actually been wanting to add the "partial release" feature for a while now. 

Instead of making releaseStake a "partialBurn and release the rest" we could make releaseStake "partial or full release" and then just call burn after.

**Comment by thegostep** [comment]
> Instead of making releaseStake a "partialBurn and release the rest" we could make releaseStake "partial or full release" and then just call burn after.

Added in https://github.com/numerai/contract/pull/80

File: contracts/NumeraireDelegateV4.sol [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+        stake.amount = 0;
+        balanceOf[_staker] = safeAdd(balanceOf[_staker], safeSubtract(originalStakeAmount, _burnAmt));
+        stake.resolved = true;
+        stake.successful = _successful;
+
+        totalSupply = safeSubtract(totalSupply, _burnAmt);
+
+        StakeReleased(_tournamentID, _roundID, _staker, _tag, _burnAmt);
+        StakeDestroyed(_tournamentID, _roundID, _staker, _tag);
```
Comment (kumikoda) [inline contracts/NumeraireDelegateV4.sol]: do we have to emit this event?

what if we just emitted the stakeReleased event?

Comment (thegostep) [reply contracts/NumeraireDelegateV4.sol]: addressed

**Comment by BryantLuu** [comment]
We still need partial release for stake selection to deplete the prize pool

**Comment by kumikoda** [comment]
> We still need partial release for stake selection to deplete the prize pool

I think that is a P2. Currently, we partial select, but full burn which is bad. With this change, we will be able to partial select and partial burn, which is better. The only "bad" thing here is that if we partially select (say 1 out of the 100 NMR) your stake, we will have to lock up the entire stake for the duration of the round even though only 1 is needed. 

In order for us to fix this, we will need to completely rethink the abstraction of stake and stake resolutions. I don't see a simple way forward here. 

@BryantLuu does this make sense with what you are thinking?

**Comment by BryantLuu** [comment]
@kumikoda 
I see that we should prioritize only having 1 stake resolution, so this makes sense.

The locking up 100 NMR for 1 NMR stake is what I'm worried about. Don't know if users will be ok with that. 

If we go that route, we should change the rules that 900NMR is the minimum prizepool and in the above case, the prize pool would be 999 NMR to fit the entire stake, but I guess that means we'd have infinite liability if there was 899 stakes then 1 staker that can do any amount.

The other option is to send the user 99 NMR so only 1 is locked up, but then we'd have to do some math with stake selection and stake to resolve correctly... jank everywhere.

**Comment by kumikoda** [comment]
@BryantLuu we already lock up 100 NMR for 1NMR selected right now. What is worse, we will burn the FULL 100 even though you only stood to actually win 1NMR x payout factor.

So what we are doing is a strict improvement.

**Comment by BryantLuu** [comment]
@kumikoda oh I didn't realize that was possible right now.  :+1: Cool. Sounds good to me!

I didn't know PARTIAL was a case... In the partial case, I thought we would only have selected the earliest ones, left out the last one completely, then raised P to deplete the prize pool which seems way more sane then making someone effectly stake 1 nmr for a chance to burn 100

**Review by kumikoda — APPROVED** [review]

File: contracts/NumeraireDelegateV4.sol [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+        return true;
+    }
+
+    // Numerai calls this function to release staked tokens when the staked predictions were successful
+    function releaseStake(address _staker, bytes32 _tag, uint256 _etherValue, uint256 _tournamentID, uint256 _roundID, bool _successful) onlyOwner stopInEmergency returns (bool ok) {
+        var round = tournaments[_tournamentID].rounds[_roundID];
+        var stake = round.stakes[_staker][_tag];
+        var burnAmount = uint128(_etherValue);
+        var releaseAmount = uint128(safeSubtract(stake.amount, burnAmount));
```
Comment (BryantLuu) [inline contracts/NumeraireDelegateV4.sol:35]: Does safeSubtract raise an error if it goes below 0?

Comment (kumikoda) [reply contracts/NumeraireDelegateV4.sol:35]: yes 
https://github.com/numerai/contract/blob/master/contracts/Safe.sol#L15

**Comment by thegostep** [comment]
This PR can be closed since the changes will me implemented in the new tournament

**Pull request closed without merging**